R/check_exercise.R

```diff
@@ -0,0 +1,407 @@
+#' This allows teachers to check that their setup and solution chunks
+#' contain valid code.
+#'
+#' @param path Path to the Markdown file containing the RMarkdown.
+#'
+#' @return TRUE or FALSE invisibly.
+#' @export
+#'
+#' @examples
+#' if (interactive()){
+#'   check_exercise("sandbox/sandbox")
+#'   check_exercise("sandbox/sandbox-exos.Rmd")
+#' }
+check_exercise <- function(
+  path
+){
+
+  path <- normalizePath(path)
+
+  # set global tutorial option which we can use as a basis for hooks
+  # (this is so we don't collide with hooks set by the user or
+  # by other packages or Rmd output formats)
+  knitr::opts_chunk$set(tutorial = TRUE)
+
+
+  # The goal of this function is to check that a given
+  # Rmd is correct, in the sense that the solution can be run
+  # using the parameters that has been provided.
+  # In other words, it tries to mimic the evaluation of an exercise
+  # where the student's input == the teacher's solution
+  #
+  # Here is  how it achieves this
+  # - parse a given Rmd
+  # - create groups of chunk, based on their label:
+  # setup, solution, and checkers
+  # - reproduce the knitr context of evaluation
+  # - for each group, run an evaluate_exercise(),
+  # using the teacher's solution as an input
+  # - If this evaluate_exercise() works, then the solution
+  # is correct
+
+  # Setting learnr options
+  learnr::tutorial_options()
+
+  # When exiting the fun, we delete all the code store
+  # inside the knitr environment
+  on.exit({
+    knitr::knit_code$restore()
+  })
+
+  # Splitting the file using knitr.
+  # This will register code inside  `knitr::knit_code`
+  # The results is a list containing all the elements from the Rmd
+  # (i.e code + title + yaml)
+  res <- knitr:::split_file(
+    xfun::read_utf8(path),
+    patterns = knitr::all_patterns$md
+  )
+
+  # Given that we are in learnr, we only want the chunks that have a label
+  # This should have the same length as knitr::knit_code$get()
+  # For some reasons the result of split_file is not the same as what
+  # knitr::knit_code$get() returns
+  usefull_chunks <- res[
+    ! vapply(
+      res,
+      function(.x) is.null(.x$params$label),
+      FUN.VALUE = logical(1)
+    )
+  ]
+
+  # Here length(knitr::knit_code$get()) == length(usefull_chunks)
+
+  # Setting the names of the chunks using the labels
+  names(usefull_chunks) <- vapply(
+    usefull_chunks,
+    function(.x) .x$params$label,
+    FUN.VALUE = character(1)
+  )
+  # Here all(names(knitr::knit_code$get()) == names(usefull_chunks))
+
+  # As the result of split_file does not contain the code, we add a code
+  # elements
+  for (i in seq_along(usefull_chunks)){
+
+    usefull_chunks[[i]]$code <- knitr::knit_code$get(
+      usefull_chunks[[i]]$params$label
+    )
+
+  }
+
+  # We extract the default setup chunk, and remove it from the list
+  setup_chunk <- usefull_chunks[["setup"]]
+  usefull_chunks[["setup"]] <- NULL
```

Comment:
Should only need to evaluate this once. Save the result into an environment

```suggestion
  setup_chunk <- usefull_chunks[["setup"]]
  usefull_chunks[["setup"]] <- NULL
  setup_chunk_envir <- eval(parse(text = setup_chunk), envir = new.env(parent = globalenv()))
```